### DISCUSSION TITLE
Sending AT Commands to radio module

### DISCUSSION THREAD

**Discussion by blueed9696** [alanbjohnston/CubeSatSim#336] [General]
Hello Alan,

I'm trying to send AT commands to radio module SU105U to get a full control of it. I'm modifying your 'program_radio' function in 'main.c ' and implementing 'write()' to send command to radio module through serial port 'dev/ttyAMA0'. 

However, it's not working as expected. What I'm assuming is the serial port may not be connected to SU105U. I'm wondering if there will be solution to send AT commands. Piece of code that I can reference for receiving data will be also helpful.

Thank you in advance.

-Jay

**Comment by alanbjohnston** [comment]
Hi Jay,

It is possible to modify the code to program the SR105U (also known as the FR_SRS_05W), U6 in the STEM Payload Beta v1.3.2 schematic.

I have used the program_radio function in main.c (beta branch) in the past, but right now, I use this python code in rpitx.py (beta branch) which operates the same:

		print("Programming FM module!\n");	
		output(pd, 1)
		output (ptt, 1)
		try:
			ser = serial.Serial("/dev/ttyAMA0", 9600)
			print(ser.portstr)
	#		uhf_string = "AT+DMOSETGROUP=0," + rx +"," + tx + ",0,3,0,0\r\n"
			uhf_string = "AT+DMOSETGROUP=0," + rx + "," + tx + "," + rxpl_value + "," + sq + "," + txpl_value + ",0\r\n"
			print(uhf_string)
			for i in range(6):
	#			ser.write(b"AT+DMOSETGROUP=0,435.0000,434.9000,0,3,0,0\r\n")
				ser.write(uhf_string.encode())
				sleep(0.1)
			ser.close()	
		except:
			print("Error in serial write")
		output(pd, 0)

I used the documentation in http://www.kh-gps.de/sr_frs.pdf although many functions don't seem to be implemented.

I also found this code, although I haven't tried it: https://github.com/sp2ong/SRFRS

The Raspberry Pi serial Transmit Data (PI_TXD in the schematic) for the Beta STEM Payload Board is connected to the RXD pin on the SR105U.  The python code above can change the transmit or receive frequency or set squelch, PL codes, etc.  I have found that many other functions documented, such as microphone level or power output do not seem to be implemented.

The TXD pin on the SR105U is not wired to the Raspberry Pi serial port PI_RXD since the Pi Zero has only one UART and we need that pin to receive the sensor data sent by the Raspberry Pi Pico processor.

Right now, the TXD pin on the SR105U is not wired to anything in the v1.3.2.4 or earlier STEM Payload PCBs.  In the next version, I'll add a PCB jumper so you could connect it to the PI_RXD pin if you wanted to.  For now, you could wire this to the Raspberry Pi PI_RXD pin (J1, GPIO pin 10) and then as long as the Pico was not plugged in or GP0 (pin 1) was set to input, you would be able to read the output from the SR105U.  I did experiment with this a while back, but I didn't reliably receive messages from the SR105U and it wasn't useful to confirm that a command was received.  Instead, I just send the same command 5 times in a row and that seems to work OK.

I'd be interested in learning if you unlock additional features and capabilities in the SR105U through your experiments!

Alan

**Reply by blueed9696** [reply]
Thank you for your kind feedback! I'll definitely share any interesting findings I discover about the radio module! I have few more questions and will be appreciated if you could reply.

- Could you share methods you've used to check if the radio module is receiving correctly?
-  Do you mean that to enable TX of SR105U, I should connect TXD pin to Pi zero RXD pin (GPIO 10)?

**Reply by alanbjohnston** [reply]
We have Command and Control working which uses the SR105U receiver.  Details on how to set it up and use it are here: https://github.com/alanbjohnston/CubeSatSim/wiki/Command-and-Control  Command and Control Carrier mode just uses the squelch pin, so you can detect a carrier.  The output of the Serial Monitor on the Pico also shows the squelch pin so you can check it there.

The DTMF/APRS mode uses demodulated audio from the SR105U through a USB sound card into the Pi Zero which runs Direwolf to detect DTMF digits or APRS packets. 

Yes, if the SR105U TXD pin 9 is wired to the Pi Zero RX data pin 10 on the GPIO, you will see the output of the SR105U.  As I said, i have seen output, but it doesn't seem to do so reliably.

Alan

**Reply by alanbjohnston** [reply]
I had one other thought about this.  The audio output from the SR105U is a very low amplitude.  I initially needed an audio amplifier circuit (such as this one https://www.amazon.com/gp/product/B09KGZ2F6D) to make it work with USB sound cards.  Then I found this USB sound card which has a high enough amplification that no extra amplifier is needed:  https://www.amazon.com/gp/product/B00IRVQ0F8 If you haven't tried this particular USB sound card, I would recommend it.

Alan
